tests/MongoDB.Driver.Tests/JsonDrivenTests/JsonDrivenWaitForEvent.cs

```diff
@@ -0,0 +1,102 @@
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using MongoDB.Bson;
+using MongoDB.Driver.Core;
+using MongoDB.Driver.Core.Events;
+using MongoDB.Driver.Core.Servers;
+
+namespace MongoDB.Driver.Tests.JsonDrivenTests
+{
+    public sealed class JsonDrivenWaitForEvent : JsonDrivenTestRunnerTest
+    {
+        private int _count;
+        private string _event;
+        private readonly EventCapturer _eventCapturer;
+
+        public JsonDrivenWaitForEvent(
+            IJsonDrivenTestRunner testRunner,
+            Dictionary<string, object> objectMap,
+            EventCapturer eventCapturer) : base(testRunner, objectMap)
+        {
+            _eventCapturer = eventCapturer;
+        }
+
+        protected override void CallMethod(CancellationToken cancellationToken)
+        {
+            Wait();
+        }
+
+        protected override Task CallMethodAsync(CancellationToken cancellationToken)
+        {
+            return Task.Run(() => Wait());
+        }
+
+        protected override void SetArgument(string name, BsonValue value)
+        {
+            switch (name)
+            {
+                case "count":
+                    _count = value.ToInt32();
+                    return;
+
+                case "event":
+                    _event = value.ToString();
+                    return;
+            }
+
+            base.SetArgument(name, value);
+        }
+
+        // private methods
+        private Func<object, bool> MapEventNameToCondition(string eventName)
+        {
+            switch (eventName)
+            {
+                case "ServerMarkedUnknownEvent":
+                    return @event =>
+                        @event is ServerDescriptionChangedEvent serverDescriptionChangedEvent &&
+                        serverDescriptionChangedEvent.NewDescription.Type == ServerType.Unknown;
+
+                case "PoolClearedEvent":
+                    return @event => @event is ConnectionPoolClearedEvent;
+
+                default:
+                    throw new Exception($"Unexpected event type {eventName}.");
+            }
+        }
+
+        private void Wait()
+        {
+            var eventCondition = MapEventNameToCondition(_event);
+            Func<IEnumerable<object>, bool> eventsConditionWithFilterByCount = (events) => events.Count(eventCondition) >= _count;
+            var notifyTask = _eventCapturer.NotifyWhen(eventsConditionWithFilterByCount);
+
+            var timeout = TimeSpan.FromSeconds(10);
+            var testFailedTimeout = Task.Delay(timeout, CancellationToken.None);
+            var index = Task.WaitAny(notifyTask, testFailedTimeout);
+            if (index != 0)
+            {
+                var triggeredEventsCount = _eventCapturer.Events.Count(eventCondition);
```

Comment:
there is a small chance for race condition issue, but I think we can live with it since it's only about final error message